### PR TITLE
Collector's Bee Ring feature

### PR DESCRIPTION
**Pull request by Cardinalstars**
Adding a bauble that simulates a Bee Collector's Jar using NBT stored on the item.

The main logic is located in ItemBeeRing.java and InventoryBeeRing.java. GUI and container are basically copy-pasted from the Collector's Bee Jar. RingHousing.java simulates an apiary just enough to get the effect to trigger.

 I'm sorry if it's coded horribly; this is my first major Minecraft feature. :)

### CODE REVIEW

File: src/main/java/magicbees/itemInventories/InventoryBeeRing.java [hunk 1]
```diff
@@ -0,0 +1,339 @@
...
+            } catch (Exception ignored) {}
+        }
+
+        for (int i = 0; i < 5; i++) {
+            String slotKey = getSlotNBTKey(i);
+            try {
+                slotNbt2.setInteger(slotKey + "i", effectData[1].getInteger(i));
+                slotNbt2.setBoolean(slotKey + "b", effectData[1].getBoolean(i));
+                slotNbt2.setFloat(slotKey + "f", effectData[1].getFloat(i));
+            } catch (Exception ignored) {}
```
Comment (OneEyeMaker) [inline src/main/java/magicbees/itemInventories/InventoryBeeRing.java:219]: Why do you need try-catch there?

File: src/main/java/magicbees/itemInventories/InventoryBeeRing.java [hunk 2]
```diff
@@ -0,0 +1,339 @@
...
+
+        if (compound.hasKey(KEY_EFFECT_1)) {
+            NBTTagCompound nbtSlots1 = compound.getCompoundTag(KEY_EFFECT_1);
+
+            for (int i = 0; i < 5; i++) {
+                String slotKey = getSlotNBTKey(i);
+
+                if (nbtSlots1.hasKey(slotKey + "i")) {
+                    int effectNBTInt = nbtSlots1.getInteger(slotKey + "i");
+                    try {
```
Comment (OneEyeMaker) [inline src/main/java/magicbees/itemInventories/InventoryBeeRing.java:241]: Why do you use try-catch? Should null check go there or what?

File: src/main/java/magicbees/tileentity/RingHousing.java [hunk 3]
```diff
@@ -0,0 +1,232 @@
...
+import forestry.api.core.IErrorLogic;
+import forestry.api.core.IErrorState;
+
+// Class that simulates a BeeHouse using both player data and info from InventoryBeeRing
+public class RingHousing implements IBeeHousing {
+
+    EntityPlayer player;
+    BiomeGenBase biome;
+    IBeeHousingInventory inventory;
+    IBeekeepingLogic beekeepingLogic;
```
Comment (OneEyeMaker) [inline src/main/java/magicbees/tileentity/RingHousing.java:32]: Express access modifiers explicitly.

**Review by OneEyeMaker — CHANGES_REQUESTED** [review]
Please, reduce amount of copy-pasted code. Either create single Container/UI/logic implementation (for both ring and jar) or create base classes with common details.

And if you need help with coding, you can ask me on Discord (`OneEyeMaker#4423`) or other devs of GTNH.

File: src/main/java/magicbees/itemInventories/InventoryBeeRing.java [hunk 4]
```diff
@@ -0,0 +1,339 @@
...
+        for (int i = 0; i < 5; i++) {
+            String slotKey = getSlotNBTKey(i);
+            try {
+                slotNbt2.setInteger(slotKey + "i", effectData[1].getInteger(i));
+                slotNbt2.setBoolean(slotKey + "b", effectData[1].getBoolean(i));
+                slotNbt2.setFloat(slotKey + "f", effectData[1].getFloat(i));
+            } catch (Exception ignored) {}
+        }
+
+        setInventorySlot(parent);
```
Comment (OneEyeMaker) [inline src/main/java/magicbees/itemInventories/InventoryBeeRing.java:222]: Please, explain why already equipped ring should be re-equipped? Do you want to update NBT?

File: src/main/java/magicbees/item/ItemBeeRing.java [hunk 5]
```diff
@@ -0,0 +1,205 @@
...
+    }
+
+    @Override
+    public boolean canUnequip(ItemStack itemStack, EntityLivingBase entityLivingBase) {
+        return true;
+    }
+
+    public static InventoryBeeRing getInventory(EntityPlayer player, ItemStack itemStack, int slot,
+            boolean baubleFlag) {
+        return new InventoryBeeRing(itemStack, player, slot, baubleFlag);
```
[the remark this answers is not in the document]
Comment (Cardinalstars) [reply src/main/java/magicbees/item/ItemBeeRing.java:203]: I'm not sure how I would make the inventory persist without making a new one every time it ticks.

File: src/main/java/magicbees/itemInventories/InventoryBeeRing.java (hunk 2 above)
Comment (Cardinalstars) [reply src/main/java/magicbees/itemInventories/InventoryBeeRing.java:241]: effectData[0]'s implementation has private fields that can't be accessed. I have no way to get the length from these, so I need to check if I can set it. Not sure how else to do this.

File: src/main/java/magicbees/itemInventories/InventoryBeeRing.java (hunk 4 above)
Comment (Cardinalstars) [reply src/main/java/magicbees/itemInventories/InventoryBeeRing.java:222]: Yes, the NBT is updated here, so the slot must be updated. I was using references earlier to just edit the slot directly, but somewhere along the way the itemStack is copied so whatever was being passed wasn't actually the itemStack in the player's inventory.

File: src/main/java/magicbees/client/gui/ContainerEffectRing.java [hunk 6]
```diff
@@ -0,0 +1,66 @@
...
+import net.minecraft.entity.player.EntityPlayer;
+import net.minecraft.inventory.Container;
+import net.minecraft.inventory.Slot;
+import net.minecraft.item.ItemStack;
+
+import cpw.mods.fml.common.registry.GameRegistry;
+
+public class ContainerEffectRing extends Container {
+
+    InventoryBeeRing IBR;
```
[the remark this answers is not in the document]
Comment (Cardinalstars) [reply src/main/java/magicbees/client/gui/ContainerEffectRing.java:14]: Fixed. I appreciate the advice.

File: src/main/java/magicbees/item/ItemBeeRing.java [hunk 7]
```diff
@@ -0,0 +1,205 @@
...
+
+            int current = bee.getHealth();
+            int max = bee.getMaxHealth();
+            IBR.setCurrentBeeHealth((current * 100) / max);
+            IBR.setCurrentBeeColour(bee.getGenome().getPrimary().getIconColour(0));
+        }
+    }
+
+    private boolean hasDrone(ItemStack itemStack, EntityPlayer player, int slot) {
+        return getInventory(player, itemStack, slot, true).getStackInSlot(DRONE_SLOT) != null;
```
[the remark this answers is not in the document]
Comment (Cardinalstars) [reply src/main/java/magicbees/item/ItemBeeRing.java:135]: Fixed. Added and InventoryBeeRing to each method that uses it so it's only created once per onWornTick

File: src/main/java/magicbees/item/ItemBeeRing.java [hunk 8]
```diff
@@ -0,0 +1,205 @@
...
+        IBR.setEffectAndInitialize(queen);
+        IBR.effectData = queen.doEffect(IBR.effectData, housingLogic);
+        IBR.writeEffectNBT();
+
+        // Crashes irregularly for some reason I can't figure out, but this is just the bee effects around you, so it's
+        // not a huge deal. It would be nice to have though
+        // I will get around to it later. (Crash has something to do with entityFX when the player is moving)
+
+        // if (player.getEntityWorld().getWorldTime() % 5 == 0) {
+        // IBR.effectData = queen.doFX(IBR.effectData, housingLogic);
```
[the remark this answers is not in the document]
Comment (Cardinalstars) [reply src/main/java/magicbees/item/ItemBeeRing.java:157]: Happened whether or not I restricted it to serverside, I'm honestly very unsure about this one.

File: src/main/java/magicbees/item/ItemBeeRing.java [hunk 9]
```diff
@@ -0,0 +1,205 @@
...
+            } else {
+                queen.writeToNBT(IBR.contents[QUEEN_SLOT].stackTagCompound);
+            }
+        } else {
+            IBR.setThrottle(IBR.throttle + 1);
+        }
+    }
+
+    private boolean hasQueen(ItemStack itemStack, EntityPlayer player, int slot) {
+        return getInventory(player, itemStack, slot, true).getStackInSlot(QUEEN_SLOT) != null;
```
[the remark this answers is not in the document]
Comment (Cardinalstars) [reply src/main/java/magicbees/item/ItemBeeRing.java:178]: Fixed. Now passes reference as a method argument.

File: src/main/java/magicbees/tileentity/RingHousing.java (hunk 3 above)
Comment (Cardinalstars) [reply src/main/java/magicbees/tileentity/RingHousing.java:32]: Fixed. Fields are now private and final.

**Review by Alastors — DISMISSED** [review]
Looks good to me, even with the code comment, I tested it out and it does indeed work

**Comment by Dream-Master** [comment]
@OneEyeMaker please can you re check pr. 
@Alastors oneEyeMake review it so not approve it until the review is done

**Comment by Alastors** [comment]
> @OneEyeMaker please can you re check pr. @Alastors oneEyeMake review it so not approve it until the review is done

Aight, the code itself looked fine and most of his requested changes were done besides a subjective "don't copy and paste code" request, so I thought it'd be fine, the code itself works so.

**Comment by Dream-Master** [comment]
@Cardinalstars please run spotlessaply
@Alastors working code is ok but code quality need to be checked as well.
@OneEyeMaker maybe give tips and tricks how Cardinalstars improve the code if necessary

**Comment by OneEyeMaker** [comment]
@Alastors I'm sorry, but your changes made code even more messy.
@Cardinalstars After more deep research of this implementation I found performance issue which I don't know how to fix right now. I'll reimplement this idea myself when I find solution. So I close this PR.

**Comment by Alastors** [comment]
I think we should just leave this in limbo until someone fixes it instead of closing this out entirely, I don't think just saying "I don't know how to fix this, I'll reimplement this myself eventually, no one else can fix this" is a good enough reason to just nuke a feature pr

**Comment by Alastors** [comment]
Either offer a new implementation as a justification to yeet this, or offer solutions to it's lone issue, of which multiple have been offered in the discord, such as it only checking like once every minute or such, or as I proposed, only one bee being allowed in it while the effect was happening, and not allowing any more until the bee was destroyed.

**Comment by chochem** [comment]
set to draft by wishes of cardinalstar and alastor

**Comment by Alastors** [comment]
I'll fix this crapshow in the morning, I'm gonna have to examine the 9 changes and figure out which broke the camel's back

**Comment by Alastors** [comment]
posted that on the wrong pr

**Comment by OneEyeMaker** [comment]
I apologize for my previous statement. My mind was overclouded yesterday.
Summary of changes:
* Simplified Container/UI/inventory code by using existing (Forestry) implementation.
* Fixed crash with particle render on incorrect side.
* Removed unnecessary setting of inventory slot.
* Allowed wearing two rings at once with different bees providing different effects.

*Probably* needs to fix:
* Ring reads inventory from NBT on every tick. Reducing tick rate N times is not a proper solution - because bee inside ring will work N times slower and live N times longer. And potion effects will update N time more rarely.
(I say "probably", because I haven't large and populate server to test.)

**Comment by Dream-Master** [comment]
@Cardinalstars is this ready for review ?

**Comment by Cardinalstars** [comment]
@Dream-Master Yes, it should now be ready. I just needed to test on dedicated server.

**Review by Alastors — APPROVED** [review]
Yeah looks fine to me, my only comment being that this needs extensive testing in large multiplayer worlds (zeta) to ensure we don't need to do more to address the possible lag that nbt checking could be, likewise,it needs a recipe, but things like this can bee handled at a later date.

**Comment by Dream-Master** [comment]
@OneEyeMaker

**Review by OneEyeMaker — APPROVED** [review]

**Pull request merged**